enhancements/machine-api/cluster-autoscaler-operator.md

```diff
@@ -0,0 +1,290 @@
+---
+title: cluster-autoscaler-operator
+authors:
+  - "@JoelSpeed"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-05-19
+last-updated: 2020-05-19
+status: implemented
+see-also:
+replaces:
+superseded-by:
+---
+
+# Cluster Autoscaler Operator
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Provide a declarative way for users to enable autoscaling of compute resources by providing the Cluster Autoscaler
+as a Deployment with opinionated default options.
+
+Provide a way for users to opt individual MachineSets into autoscaling and to set boundaries for the size of the
+MachineSets that they wish to autoscale.
+
+## Motivation
+
+### Goals
+
+- Enable users to create Cluster Autoscaler with opinionated defaults
+- Enable users to autoscale MachineSets
+- Enable users to set a maximum size of their cluster
+- Enable users to set maximum and minimum sizes for each of their MachineSets
+
+### Non-Goals
+
+- Integrate the Cluster Autoscaler with Machine API (This will be a separate proposal)
+- Enable autoscaling by default for new OpenShift clusters
+- Provide options for autoscaling across multiple MachineSets
+
+## Proposal
+
+A new controller will be developed to manage the lifecycle of the Cluster Autoscaler application and to manage
+assigning size limits to MachineSets to enable the Cluster Autoscaler to autoscale desired MachineSets.
+
+### User Stories
+
+#### Story 1
+
+As a user of OpenShift, it should be easy to enable autoscaling of compute resources so that my cluster will respond to
+increased usage without manual intervention.
+
+#### Story 2
+
+As a user of OpenShift, I should be able to specify which groups of Machines should be autoscaled and which should not.
+
+#### Story 3
+
+As a user of OpenShift, I should be able to limit the size of my cluster when it is autoscaling so that I do not overload
+the Control Plane and so that I can cap the costs of my cluster.
+
+#### Story 4
+
+As a user of OpenShift, I should be able to specify a maximum and minimum size for any group of Machines being scaled
+so that I can ensure I have a minimum capacity available and that I do not over commit to a single pool of Machines.
+
```

Comment:
I would like to know if you consider the upgrade use-case in-scope for this.  For example, rebooting nodes to upgrade them causes a reduction in total cluster capacity.  Should this component be tied into our upgrade flow to maintain steady capacity during upgrades.

Reply:
> For example, rebooting nodes to upgrade them causes a reduction in total cluster capacity...

Rolling machine reboots are not specific to updates.  For example, they also occur when the MCO rolls a new MachineConfig out to a pool.  Would be nice if the MCO could coordinate with any enabled autoscaler so that folks could ask to surge through a rolling reboot instead of just scaling down a node or two at a time while rolling out a reboot.

Reply:
This document is meant to capture the current state of this. It was implemented before the enhancements repo existed so I'm just trying to backfill the proposal as it were.

We don't have this support at present, and I'm not sure this is part of the Cluster-Autoscaler-Operator scope anyway, this would be better suited to a discussion on the Cluster Autoscaler proposal (which I have yet to write) wouldn't it?

Reply:
I didn't realize folks are backfilling the enhancements repo.  That's confusing.

Reply:
The idea is that we should have a document detailing what the current state is, so that should we want to make future enhancements, we can refer back to the current implementation so that reviewers have a full context for the future enhancements and the impact they will have on the existing state of the projects